(Major) Refactor

aarondill opened the pull request:
This PR refactors `download.js` for readability, maintainability, and user experience.
The intent of this PR is *not* to change behavior, although there are a few changes (see below), but rather to improve the codebase maintainability and readability.

Note: this builds on top of #3. This may need to be rebased once it's merged.

### Changes made to behavior:
1. Don't check if fd is installed. See #3.
2. Shell command usage is now safer. This shouldn't be noticed in most cases, but it helps avoid edge cases.
3. platform support is checked at the beginning of `checkRequirements`, before required command checking. This helps unsupported users to determine the lack of support faster, without requiring installation of commands just to be met with a "Not supported" message.
4. The github response must include a status code of 200, or the script aborts
5. if there is not `assets` array in the returned json, an error is thrown manually, which allows for easier debugging and user experiance.
6. On thrown errors, `process.exitcode` is set to 1 and `error.message` is output if present, else the full error. This removes the stack trace, and thusly makes errors easier to understand for the user.
7. If there is an error during the request to GitHub (ie, internet issues), a message is output to the user reminding them to check their internet.
8. The commands run to download and install `fd` now output and accept input from the terminal. This shouldn't been noticed except in edge cases, as all the commands should be quiet.
9. If a command fails during installation, the installation is aborted **and the user is informed**

aarondill committed:
Refactor commandExists

This removes some of the callback hell by using promises

aarondill committed:
Don't check if fd is installed

Checking has several issues:
1. On updates, `fd` is already installed, resulting in a failure
2. The package.json includes `"bin": { "fd": "dist/fd" }`, which gets installed before the script runs, resulting in the script *always* failing.
3. There is *no* reason to deny multiple instances of `fd` on a system. Perhaps I have `fd` installed through `apt`, but *also* want to have the latest version through `npm`. This commit (if it worked correctly) blocks this use case.

aarondill committed:
formatting

Ran prettier

aarondill committed:
check platform support at script start

aarondill committed:
clean up shell command invocations

This makes it more readable, handles spaces in file names, ensure absolute paths, handles paths starting with `-`, etc

aarondill committed:
refactor checkRequirements

This makes the oneliner that finds missing commands easier to understand.
It also uses .forEach instead of .filter, as the resulting array is not needed.

aarondill committed:
refactor main

This moves the options to a variable to increase readability
It also checks the res.statusCode to ensure it is 200
Finally, it does a small amount of data validation to ensure obj.assets exists and is an array

aarondill committed:
nicer error handling

This outputs error messages if possible to the user and sets the program to exit cleanly with a code of 1

aarondill committed:
fix nicer error handling

This wraps the main invocation in an async iife to allow catching of rejected promises

aarondill committed:
fix command_exists

Apparently the order of .then() and .catch() is important

aarondill committed:
output message on request failure.

aarondill committed:
fix spawnSync usage

`spawnSync` takes an array of arguments as it's second argument and a object for options as it's third.
It also doesn't throw.

the new `runCommand` utility will throw if the command fails, and will allow the command to output to the terminal (for debugging purposes)

aarondill committed:
move platform check to `checkRequirements`

This should still run first, but now it's error should be handled by the try catch at the bottom. Additionally, it makes sense in the section for checking requirements

pablopunk commented on download.js:
```diff
@@ -35,75 +34,101 @@ const chooseAsset = (assets) => {
 }
 
-checkRequirements().then(main);
+(async function () {
+  try {
+    await checkRequirements();
+    main();
+  } catch (e) {
+    process.exitCode = 1;
+    if (e instanceof Error) console.error(e.message);
+    else console.error(e);
+  }
+})();
```
I made the `main` function to avoid this, can you rename the function to `installFd` or something like that and then change this one to `main`?

```suggestion
async function main() {
  await checkRequirements();
  installFd();
}

main()
  .catch(e => {
    process.exitCode = 1;
    if (e instanceof Error) console.error(e.message);
    else console.error(e);
  })
```

aarondill replied:
This is added in [d2a01f3](https://github.com/pablopunk/fd-find/pull/4/commits/d2a01f3ac827c35fde204b6f09024546349c00aa)

pablopunk commented on download.js:
```diff
@@ -35,75 +34,101 @@ const chooseAsset = (assets) => {
 async function checkRequirements() {
+  if (!platformFiles.hasOwnProperty(platform)) {
+    throw new Error(
+      `platform '${platform}' is not yet supported by this script`
```
may be worth to include some info here
```suggestion
      `Platform '${platform}' is not yet supported by this tool. Feel free to open an issue on https://github.com/pablopunk/fd-find/issues`
```

aarondill committed:
Add additional information to the error for unsupported platforms

Co-authored-by: Pablo P Varela <pablo@pablopunk.com>

aarondill committed:
rename main to installFd and create a new main

This allows .catch instead of an IIFE. It also makes the code slightly cleaner

Co-authored-by: Pablo P Varela <pablo@pablopunk.com>

pablopunk reviewed (approved):
Awesome blossom 🙌🏻 🌼

pablopunk commented:
🚀 Released in [1.2.0](https://github.com/pablopunk/fd-find/releases/tag/1.2.0)
